### PR TITLE
xtest: regression 41xx target Secure Key Services tests

### PR DESCRIPTION
**Pull request by etienne-lms**
Introduce regression test 41xx for PKCS#11 and SKS TA testing.

**edited: this P-R depends on RFCs https://github.com/OP-TEE/optee_os/pull/2732 and https://github.com/OP-TEE/optee_client/pull/138.**

### CODE REVIEW

File: host/xtest/Makefile [hunk 1]
```diff
@@ -59,6 +59,15 @@ ifeq ($(CFG_SECURE_DATA_PATH),y)
 srcs += sdp_basic.c
 endif
 
+CFG_SECURE_KEY_SERVICES ?= y
```
Comment (jenswi-linaro) [inline host/xtest/Makefile]: Why do we assign this here?

File: host/xtest/CMakeLists.txt [hunk 2]
```diff
@@ -24,6 +24,7 @@ set (SRC
 	benchmark_2000.c
 	regression_1000.c
 	regression_4000.c
+	regression_4100.c
```
Comment (jenswi-linaro) [inline host/xtest/CMakeLists.txt]: No conditional on `CFG_SECURE_KEY_SERVICES` here?

File: host/xtest/regression_4100.c [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+static void xtest_tee_test_4101(ADBG_Case_t *c)
+{
+	CK_RV rv;
+
+	rv = C_Initialize(NULL);
+	if (!ADBG_EXPECT_COMPARE_UNSIGNED(c, rv, ==, CKR_OK))
+		return;
+
+	rv = C_Finalize(NULL);
+	ADBG_EXPECT_COMPARE_UNSIGNED(c, rv, ==, CKR_OK);
```
Comment (jenswi-linaro) [inline host/xtest/regression_4100.c]: `if (` missing?

File: host/xtest/regression_4100.c [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+#include "xtest_test.h"
+#include "xtest_helpers.h"
+
+static void xtest_tee_test_4101(ADBG_Case_t *c)
+{
+	CK_RV rv;
+
+	rv = C_Initialize(NULL);
+	if (!ADBG_EXPECT_COMPARE_UNSIGNED(c, rv, ==, CKR_OK))
```
Comment (jenswi-linaro) [inline host/xtest/regression_4100.c]: It seems we're going to test a `CK_RV` type quite often in this suite. Consider supplying special macros as for `TEEC_Result`.

File: host/xtest/CMakeLists.txt (hunk 2 above)
Comment (etienne-lms) [reply host/xtest/CMakeLists.txt]: yes. thanks.

File: host/xtest/Makefile (hunk 1 above)
Comment (etienne-lms) [reply host/xtest/Makefile]: :)

File: host/xtest/regression_4100.c (hunk 4 above)
Comment (etienne-lms) [reply host/xtest/regression_4100.c]: Ok, good idea.

File: host/xtest/regression_4100.c (hunk 3 above)
Comment (etienne-lms) [reply host/xtest/regression_4100.c]: :), nice test!
thanks

**Comment by etienne-lms** [comment]
comments addressed.
need to rebase...

**Comment by etienne-lms** [comment]
rebased.

**Comment by etienne-lms** [comment]
Updated according to recent changes (library renaming) in https://github.com/OP-TEE/optee_client/pull/121.

**Comment by etienne-lms** [comment]
Rebase on latest master. 
PKCS#11/SKS tests have moved to  regression_42xx since regression_41xx are already assigned.

This P-R is still a first step for **libsks** to propose a PKCS#11 client API. This P-R introduces a regression test series. The sole test done here is calling the library initialization;

File: host/xtest/Makefile [hunk 5]
```diff
@@ -77,6 +77,14 @@ ifeq ($(CFG_SECURE_DATA_PATH),y)
 srcs += sdp_basic.c
 endif
 
+ifeq ($(CFG_SECURE_KEY_SERVICES),y)
+srcs += regression_4200.c
+OPTEE_SKS_HEADERS ?= ../../../optee_client/libsks/include
```
Comment (jenswi-linaro) [inline host/xtest/Makefile]: Please don't reach outside of this git into some other git for internal .h files. Whatever .h files are needed should be installed somewhere where we can find them.

Comment (etienne-lms) [reply host/xtest/Makefile]: ok. I will remove this line and `CFLAGS += -I$(OPTEE_SKS_HEADERS)` below.
The lib headers are already exported in `$(OPTEE_CLIENT_EXPORT)/include`.

**Comment by jenswi-linaro** [comment]
Please replace the SKS acronym and Secure Key Services with pkcs11 since that's what it's about.
Add this as a separate suit in xtest, pkcs11 for instance.
+1 to a test for `SKS_CMD_PING`.

**Comment by etienne-lms** [comment]
With this latest update, pkcs11 TA tests are no more in the regression tests and the TA is no more default embedded within Qemu OP-TEE core.
To test the pkcs11 TA:
* https://github.com/OP-TEE/optee_os/pull/2732 + https://github.com/OP-TEE/optee_client/pull/138 + this P-R #281 
* Build from `make all CFG_PKCS11_TA=y CFG_IN_TREE_EARLY_TAS=pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee`
* On target: `xtest -t pkcs11`
Expected output trace:
```
######################################################
#
# pkcs11
#
######################################################
* pkcs11_1000 Initialize and close Cryptoki library
  pkcs11_1000 OK
+-----------------------------------------------------
Result of testsuite pkcs11:
pkcs11_1000 OK
+-----------------------------------------------------
5 subtests of which 0 failed
1 test case of which 0 failed
0 test cases were skipped
TEE test application done!
```

**Comment by etienne-lms** [comment]
For info, IBART fails because it needs `libckteec` from OP-TEE/optee_client#138:
```
 32693:  /home/optee/devel/reference/toolchains83_201903/aarch64/bin/../lib/gcc/aarch64-linux-gnu/8.3.0/../../../../aarch64-linux-gnu/bin/ld: cannot find -lckteec
```

**Comment by jbech-linaro** [comment]
> For info, IBART fails because it needs `libckteec` from [OP-TEE/optee_client#138](https://github.com/OP-TEE/optee_client/pull/138):

Let me know if I need to update the build config in IBART ... add flags etc.

**Comment by etienne-lms** [comment]
> Let me know if I need to update the build config in IBART ... add flags etc.

Do you want IBART always run the pkcs11 tests ? now and in the future?

**Comment by etienne-lms** [comment]
For info, the pkcs11 TA gets pinged when client opens (initializes) the cryptoki library, i.e. from test `pkcs11_1000` in the P-R.

**Comment by etienne-lms** [comment]
R-B tag applied.
Depends on https://github.com/OP-TEE/optee_os/pull/2733 and https://github.com/OP-TEE/optee_client/pull/138.

Maybe add the pkcs11 to IBART.
Need build with `CFG_PKCS11_TA=y` and embedding the TA image, i.e. adding TA to early TAs: `CFG_IN_TREE_EARLY_TAS="avb/023f8f1a-292a-432b-8fc4-de8471358067 pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee"`.

**Comment by jbech-linaro** [comment]
> Maybe add the pkcs11 to IBART.

Maybe? Give me a 'yes' or 'no' instead and I'll do what you think makes sense :)

**Comment by jforissier** [comment]
> > Maybe add the pkcs11 to IBART.
> 
> Maybe? Give me a 'yes' or 'no' instead and I'll do what you think makes sense :)

:smile: 

I have just merged the optee_client and optee_os PRs, so we're good from a dependency point of view. And I agree it would be nice to have this checked by IBART.

**Comment by etienne-lms** [comment]
@jbech-linaro, yes, add to IBART please.

**Comment by jbech-linaro** [comment]
> @jbech-linaro, yes, add to IBART please.

That was clear! :smile: 

So, I'll add
```bash
make ... CFG_PKCS11_TA=y CFG_IN_TREE_EARLY_TAS="avb/023f8f1a-292a-432b-8fc4-de8471358067 pkcs11/fd02c9da-306c-48c7-a49c-bbd827ae86ee"
```
and we're good to go?

**Comment by etienne-lms** [comment]
Should be, yes.
You should see a trace `pkcs11_1001 OK` in `xtest` logs.

**Comment by etienne-lms** [comment]
thanks

**Pull request merged**